### PR TITLE
Integrate the SCP nomination protocol

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
This requires ~https://github.com/bpfkorea/agora/pull/364~.

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
I'll split up the 3 non-controversial commits into another PR as they're ready to go.

**Comment by AndrejMitrovic** [comment]
The `Nominator` class doesn't have to exist, it can be rolled into `Driver`. Initially I split them up thinking that the nominating part and the driver part (which actually sends messages) would benefit from being separate, but there's not much additional functionality the `Nominator` has.

**Comment by AndrejMitrovic** [comment]
Updated with latest fixes. The test-suite passes now (if the LocalRest change is included).

The big SCP commit needs to be further broken down. And I'll focus on writing additional tests, probably porting some test-cases from stellar-core.

**Comment by AndrejMitrovic** [comment]
I've got a race-condition in my code. The following triggers a segfault, as it tries to query the database after it's been shut down:

```diff
diff --git a/source/agora/consensus/data/UTXOSet.d b/source/agora/consensus/data/UTXOSet.d
index 8c8b06d9..de5aeb94 100644
--- a/source/agora/consensus/data/UTXOSet.d
+++ b/source/agora/consensus/data/UTXOSet.d
@@ -105,6 +105,7 @@ public class UTXOSet
     public void shutdown ()
     {
         this.utxo_db.shutdown();
+        this.utxo_db = null;
     }

     /***************************************************************************
diff --git a/source/agora/utils/Log.d b/source/agora/utils/Log.d
index 5b043acc..5c6e2281 100644
--- a/source/agora/utils/Log.d
+++ b/source/agora/utils/Log.d
@@ -42,12 +42,8 @@ public template AddLogger (string moduleName = __MODULE__)
 /// Convenience alias
 public alias LogLevel = Level;

-version (unittest) {}
-else
+/// Initialize the logger
+static this ()
 {
-    /// Initialize the logger
-    static this ()
-    {
-        Log.defaultConfig();
-    }
+    Log.defaultConfig();
 }
```

**Comment by AndrejMitrovic** [comment]
Updated: Now we are nominating transaction sets, rather than whole blocks. This will allow implementing tx set merging rules in the future, for example: take first 8 sorted txs out of the union of all the proposed tx sets.

**Comment by AndrejMitrovic** [comment]
This code includes https://github.com/bpfkorea/agora/pull/364, but I'm thinking of a way to avoid this. Maybe by implementing `setupTimer()` on the C++ side and avoiding the need to change the function signatures in https://github.com/bpfkorea/agora/pull/364/files.

**Comment by Geod24** [comment]
```
Linking...
.dub/build/unittest-unittest-cov-linux.posix-x86_64-dmd_2088-60157B2BF871D74430F234139CD56BE4/agora-unittests.o: In function `_D5agora9consensus8protocol9NominatorQk22nominateTransactionSetMFNemSQCp6common3Set__TQhTSQDkQDh4data11TransactionQnZQBmQCaZv':
/home/travis/build/bpfkorea/agora/source/agora/consensus/protocol/Nominator.d:130: undefined reference to `stellar::SCP::nominate(unsigned long, xdr::xvector<unsigned char, 4294967292u> const&, xdr::xvector<unsigned char, 4294967292u> const)'
.dub/build/unittest-unittest-cov-linux.posix-x86_64-dmd_2088-60157B2BF871D74430F234139CD56BE4/agora-unittests.o:(.data._D5agora9consensus8protocol9NominatorQk6__vtblZ+0x58): undefined reference to `stellar::SCPDriver::computeValueHash(unsigned long, xdr::xvector<unsigned char, 4294967292u> const&, int, xdr::xvector<unsigned char, 4294967292u> const)'
```

**Comment by AndrejMitrovic** [comment]
2.088.1 linker failure due to https://issues.dlang.org/show_bug.cgi?id=20336

We can't revert to 2.088.0 because it also has a mangling problem which 2.088.1 solved.

**Comment by AndrejMitrovic** [comment]
I just hard a random segfault (DMD) and test failure (LDC) locally, and it doesn't always reproduce. I have to debug this.

**Comment by AndrejMitrovic** [comment]
@Geod24 https://github.com/dlang/dmd/pull/10477 fixed the linking issue, but when are they going to release 2.088.2 with the fix?

**Comment by Geod24** [comment]
It's going to be part of 2.089.1, as it missed the deadline for 2.089.0 ... https://github.com/dlang/dmd/pull/10527#issuecomment-549096588

**Comment by AndrejMitrovic** [comment]
Resolved conflicts. It will still fail due to linking issues that will hopefully be resolved soon once the new DMD comes out.

**Comment by Geod24** [comment]
Rebase ?

**Comment by AndrejMitrovic** [comment]
> Rebase ?

Done. But it will still fail to link on ldc until they make a new release based on the bugfixes in 2.089.1.

**Comment by Geod24** [comment]
So we need to drop LDC 1.18.0 for this to move forward.
In order to do that, we need to update the LDC package on Alpine linux.
On it: https://gitlab.alpinelinux.org/alpine/aports/merge_requests/2703

**Comment by Geod24** [comment]
The fix to the package is done, we can drop LDC 1.18.0 from the CI

**Comment by Geod24** [comment]
https://github.com/bpfkorea/agora/pull/463

**Comment by AndrejMitrovic** [comment]
Rebased.

**Comment by AndrejMitrovic** [comment]
There is a random stack overflow on LDC, restarting the build. SCP does use more stack space, so it might be that it's caused by the same thing as https://github.com/dlang/dub/issues/1812#issuecomment-568981831

File: source/agora/consensus/protocol/Nominator.d [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+            if (timerID !in this.timers)
+                return;
+
+            callCPPDelegate(cb);
+        });
+    }
+}
+
+/// workaround for logger not being nothrow
+private void logCall (T...)(lazy T args, string func = __FUNCTION__) nothrow
```
Comment (Geod24) [inline source/agora/consensus/protocol/Nominator.d]: Can we make a `nothrow` wrapper in Agora an call it a day ?

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]: Like a wrapper around `Logger` with an `alias logger this` inside?

File: source/agora/consensus/protocol/Nominator.d [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+
+        this.timers.put(timerID);
+
+        this.taskman.runTask(
+        {
+            this.taskman.wait(timeout.msecs);
+            if (timerID !in this.timers)
+                return;
+
+            callCPPDelegate(cb);
```
Comment (Geod24) [inline source/agora/consensus/protocol/Nominator.d]: ~You never remove the delegate from `this.timers` though, or did I miss something ?~

So this function is called both to setup and remove timers, it seems. Could you document it a bit ?

File: source/scpd/Cpp.d [hunk 3]
```diff
@@ -73,6 +74,10 @@ extern(C++, (StdNS!())) {
 /// C++ support for foreach
 extern(C++) private int cpp_set_foreach(T)(void* set, void* ctx, void* cb);
 
+// todo fix: SCPQuorumSet should be const, but triggers a wild vibe.d error
```
Comment (Geod24) [inline source/scpd/Cpp.d]: Can you add more details about the error ?

File: source/scpd/Cpp.d [hunk 4]
```diff
@@ -27,6 +27,7 @@
 
 module scpd.Cpp;
 
+import scpd.types.Stellar_SCP;
```
Comment (Geod24) [inline source/scpd/Cpp.d]: This should actually be in `scpd.Utils`.
The point of this module is that it provides "C++" wrapper, as in, not specific to Stellar, and that ultimately we'll be able to remove it.

File: source/agora/consensus/protocol/Nominator.d [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+            quorum_set = the quorum set of this node
+
+    ***************************************************************************/
+
+    public this (KeyPair key_pair, Ledger ledger,
+        TaskManager taskman, NetworkClient[PublicKey] peers,
+        SCPQuorumSet quorum_set)
+    {
+        this.key_pair = key_pair;
+        import scpd.types.Stellar_types : StellarHash = Hash;
```
Comment (Geod24) [inline source/agora/consensus/protocol/Nominator.d]: This also exists at module scope

File: source/agora/consensus/protocol/Nominator.d (hunk 1 above)
Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]: https://github.com/bpfkorea/agora/pull/464

**Comment by AndrejMitrovic** [comment]
> Can we make a nothrow wrapper in Agora an call it a day ?

Are you still reviewing? =)

It's also green, for the first time ever. Yay.

File: source/agora/consensus/protocol/Nominator.d [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+
+        assert(0);  // should not reach here
+    }
+
+    ///
+    public override void setupTimer (ulong slotIndex, int timerID,
+        milliseconds timeout, CPPDelegate!SCPCallback* cb)
+    {
+        scope (failure) assert(0);
+        logCall("{} {}", this.key_pair.address.prettify);
```
Comment (Geod24) [inline source/agora/consensus/protocol/Nominator.d]: That's just disgusting

File: source/agora/node/Ledger.d [hunk 7]
```diff
@@ -91,6 +99,43 @@ public class Ledger
...
+    /***************************************************************************
+
+        Set the transaction set nominator callback.
+
+        Params:
+            nominateTxSet = the nominator callback
+
+    ***************************************************************************/
+
+    public void setNominator ( NominatorDg nominateTxSet ) @safe
```
Comment (Geod24) [inline source/agora/node/Ledger.d]: Extra whitespace after parenthesis

File: source/agora/node/Ledger.d [hunk 8]
```diff
@@ -49,6 +54,9 @@ public class Ledger
     /// UTXO set
     private UTXOSet utxo_set;
 
+    /// Callback when a transaction set is ready to be nominated
+    public NominatorDg nominateTxSet;
```
Comment (Geod24) [inline source/agora/node/Ledger.d]: Why not just make it a method and pass its address when needing a delegate ?

File: source/agora/node/Node.d [hunk 9]
```diff
@@ -107,10 +114,47 @@ public class Node : API
...
     public void start ()
     {
+        import scpd.scp.QuorumSetUtils;
         log.info("Doing network discovery..");
-        this.network.discover();
-
+        auto peers = this.network.discover();
         this.network.retrieveLatestBlocks(this.ledger);
+
+        if (this.config.node.is_validator)
```
Comment (Geod24) [inline source/agora/node/Node.d]:
```D
if (!this.config.node.is_validator)
    return;
```

File: source/agora/test/Base.d [hunk 10]
```diff
@@ -485,9 +486,12 @@ public class TestNode : Node, TestAPI
 /// Describes a network topology for testing purpose
 public enum NetworkTopology
 {
-    /// 4 nodes which all know about each other. Figure 9 in the SCP paper.
+    /// A number nodes which all know about each other. Figure 9 in the SCP paper.
```
Comment (Geod24) [inline source/agora/test/Base.d]: A number of... ?

File: source/agora/test/Base.d [hunk 11]
```diff
@@ -559,55 +564,98 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager)
...
+    case NetworkTopology.Cyclic:
+        node_configs = iota(nodes).map!(_ => makeNodeConfig()).array;
+        node_configs.each!((ref conf) => conf.min_listeners = 1);
+        configs = iota(nodes)
+            .map!(idx => makeCyclicConfig(idx, node_configs[idx], node_configs))
+                .array;
 
+        break;
+
+    case NetworkTopology.SinglePoF:
```
Comment (Geod24) [inline source/agora/test/Base.d]: Why the move ?

**Review by Geod24 — COMMENTED** [review]
First round of review complete. I need a second round for sure though. In the morning. After a coffee. ~And some speed~

File: source/agora/test/Base.d (hunk 11 above)
Comment (AndrejMitrovic) [reply source/agora/test/Base.d]: The diff seems to make it harder to see, but previously we only created nodes for the `NetworkTopology.Simple` config, now we create configs for `NetworkTopology.Simple`, `NetworkTopology.OneNonValidator`, and `NetworkTopology.Cyclic`, and we move the instantiation of the nodes after the configs were built (so we don't duplicate node creation code).

File: source/scpd/Cpp.d (hunk 3 above)
Comment (AndrejMitrovic) [reply source/scpd/Cpp.d]: It seems to be gone, and now const works. I don't remember what the error was. :/

File: source/agora/node/Ledger.d (hunk 8 above)
Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d]: This was accidentally public, it should be private. I'll fix it.

**Comment by AndrejMitrovic** [comment]
Rebased, fixed comments, documented, and removed the `logCall` stuff too.

File: source/scpd/types/Utils.d [hunk 12]
```diff
@@ -18,6 +18,10 @@ import scpd.types.Stellar_SCP;
 import scpd.types.Stellar_types;
 import scpd.types.XDRBase;
 
+// todo fix: SCPQuorumSet should be const, but triggers a wild vibe.d error
```
Comment (Geod24) [inline source/scpd/types/Utils.d]: What's the error again ? Is it RT or CT ?

File: source/agora/node/Ledger.d (hunk 8 above)
Comment (Geod24) [reply source/agora/node/Ledger.d]: I mean it could still be a method instead of a delegate no?

File: source/agora/test/Base.d (hunk 11 above)
Comment (Geod24) [reply source/agora/test/Base.d]: I meant that L608 (`case NetworkTopology.SinglePoF:`) has just been reordered and is still doing the same (assert(0)), but not a big deal.

File: source/scpd/types/Utils.d [hunk 13]
```diff
@@ -18,9 +18,8 @@ import scpd.types.Stellar_SCP;
 import scpd.types.Stellar_types;
 import scpd.types.XDRBase;
 
-// todo fix: SCPQuorumSet should be const, but triggers a wild vibe.d error
 extern(C++) public shared_ptr!SCPQuorumSet makeSharedSCPQuorumSet (
-    ref SCPQuorumSet);
+    ref const(SCPQuorumSet));
```
Comment (Geod24) [inline source/scpd/types/Utils.d:22]: Seems to belong to a previous commit

File: source/agora/test/GossipProtocol.d [hunk 14]
```diff
@@ -51,22 +51,29 @@ unittest
...
-        2.seconds,
-        format("Node %s has block height %s. Expected: %s",
-        idx,
-        node.getBlockHeight().to!string,
-        1)));
+    version (none)
+    {
+        // todo: race condition: when the 8th transaction is added to one node,
+        // it begins the nomination protocol before the transaction had the
+        // chance to be propagated to other nodes
```
Comment (Geod24) [inline source/agora/test/GossipProtocol.d]: That's a pretty big bug heh

**Comment by Geod24** [comment]
Looks good except for the `version (none)`. What's the plan for it ?

File: source/agora/test/GossipProtocol.d (hunk 14 above)
Comment (AndrejMitrovic) [reply source/agora/test/GossipProtocol.d]: Yeah so, Stellar nomination is based on a timer, but we don't want to physically wait 10 minutes in the **test-suite** before each nomination round starts. So I've temporarily implemented it in a way that the nomination starts as soon as there's enough transactions in the pool. 

This is not ideal. We probably want to use, ehh, a virtual timer of sorts instead? :>

File: source/scpd/types/Utils.d (hunk 12 above)
Comment (AndrejMitrovic) [reply source/scpd/types/Utils.d]: The issue is I can't reproduce it anymore. /shrug

File: source/agora/test/Base.d (hunk 11 above)
Comment (AndrejMitrovic) [reply source/agora/test/Base.d]: Oh good eye! I'll fix that up.

**Comment by AndrejMitrovic** [comment]
> Looks good except for the version (none). What's the plan for it ?

I just fixed the "timing issue" bug now. But the BanManager test is still versioned out. I think I can fix it too, today.

**Comment by AndrejMitrovic** [comment]
Fixed the BanManager test. I just added a non-validator node and tested banning behavior with that one.

**Comment by AndrejMitrovic** [comment]
Fixed a bug just now. I used `%s` by mistake and it threw an exception.

**Comment by Geod24** [comment]
Needs a rebase now

**Comment by AndrejMitrovic** [comment]
@Geod24 ready to go.

File: source/agora/test/BanManager.d [hunk 15]
```diff
@@ -25,30 +25,34 @@ import agora.consensus.Genesis;
 import agora.test.Base;
 
 /// test node banning after putTransaction fails a number of times
+/// note: temporarily disabled as node banning needs to be rethinked with regards
+/// to the node being in another node's quorum set
```
Comment (Geod24) [inline source/agora/test/BanManager.d]: Not valid anymore

File: source/agora/test/GossipProtocol.d [hunk 16]
```diff
@@ -51,6 +51,7 @@ unittest
     auto send_txs = txs[0..$-1];
     // send it to tx to node
     send_txs.each!(tx => node_1.putTransaction(tx));
+
```
Comment (Geod24) [inline source/agora/test/GossipProtocol.d]: ?

**Review by Geod24 — APPROVED** [review]
LGTM just 2 nits

**Comment by AndrejMitrovic** [comment]
Yep, sorry for missing these. Fixed now.

**Comment by Geod24** [comment]
💣💥

**Pull request merged**